File: packages/core/src/config/parse.ts
/**
 * Hand-rolled validation (zero runtime deps). Collects ALL violations and
 * reports them with JSON-path style locations so both SDKs can emit
 * identical messages. Keep in lockstep with the Python validator.
 *
 * Supports `${ENV_VAR}` interpolation in string values. Resolved against
 * the provided env map or `process.env` when available.
 */

import { ConfigError } from "../errors.js";
import { knownProviderIds } from "../providers/registry.js";
import { getPreset } from "../providers/presets.js";
import type { ModelRoute, RouterConfig } from "./schema.js";

const ENV_PATTERN = /\$\{([A-Za-z_][A-Za-z0-9_]*)\}/g;

/** Known fields at each config level; anything else is a typo and rejected. */
const TOP_LEVEL_FIELDS: ReadonlySet<string> = new Set(["routes", "strategy"]);
const ROUTE_FIELDS: ReadonlySet<string> = new Set([
  "id",
  "provider",
  "model",
  "apiKey",
  "apiKeys",
  "baseUrl",
  "apiVersion",
  "region",
  "project",
  "headers",
  "maxRetries",
  "timeoutMs",
  "streamIdleTimeoutMs",
  "limit",
  "budget",
  "weight",
]);
const LIMIT_FIELDS: ReadonlySet<string> = new Set(["rpm", "tpm"]);
const BUDGET_FIELDS: ReadonlySet<string> = new Set(["usd", "windowMs"]);
const STRATEGIES: ReadonlySet<string> = new Set([
  "fallback",
  "round-robin",
  "weighted",
  "least-latency",
]);

/**
 * Recursively resolve `${VAR}` patterns in string values.
 * Missing vars throw a clear error.
 */
function interpolateEnv(value: unknown, env: Record<string, string | undefined>, path: string): unknown {
  if (typeof value !== "string") return value;
  if (!value.includes("${")) return value;
  return value.replace(ENV_PATTERN, (_, varName: string) => {
    const resolved = env[varName];
    if (resolved === undefined) {
      throw new ConfigError(`${path}: environment variable "${varName}" is not set`);
    }
    return resolved;
  });
}

function interpolateDeep(obj: unknown, env: Record<string, string | undefined>, path: string): unknown {
  if (typeof obj === "string") return interpolateEnv(obj, env, path);
  if (Array.isArray(obj)) return obj.map((item, i) => interpolateDeep(item, env, `${path}[${i}]`));
  if (obj !== null && typeof obj === "object") {
    const result: Record<string, unknown> = {};
    for (const [key, val] of Object.entries(obj as Record<string, unknown>)) {
      result[key] = interpolateDeep(val, env, `${path}.${key}`);
    }
    return result;
  }
  return obj;
}

function isObject(v: unknown): v is Record<string, unknown> {
  return typeof v === "object" && v !== null && !Array.isArray(v);
}

function positiveInt(v: unknown): boolean {
  return typeof v === "number" && Number.isInteger(v) && v > 0;
}

export function parseConfig(input: unknown, env?: Record<string, string | undefined>): RouterConfig {
  // Resolve ${VAR} patterns against env or process.env.
  const resolvedEnv = env ?? (typeof process !== "undefined" ? process.env : {});
  const interpolated = interpolateDeep(input, resolvedEnv, "config");

  const errors: string[] = [];

  if (!isObject(interpolated)) {
    throw new ConfigError("config: expected an object");
  }

  if (!Array.isArray(interpolated.routes)) {
    throw new ConfigError("config.routes: expected an array");
  }
  if (interpolated.routes.length === 0) {
    throw new ConfigError("config.routes: must not be empty");
  }

  for (const key of Object.keys(interpolated)) {
    if (!TOP_LEVEL_FIELDS.has(key)) {
      errors.push(`config.${key}: unknown field (known: routes, strategy)`);
    }
  }

  const strategy = interpolated.strategy;
  if (strategy !== undefined && (typeof strategy !== "string" || !STRATEGIES.has(strategy))) {
    errors.push(
      'config.strategy: must be "fallback", "round-robin", "weighted", or "least-latency"',
    );
  }

  const seenIds = new Set<string>();
  const routes: ModelRoute[] = [];

  interpolated.routes.forEach((raw, i) => {
    const at = `config.routes[${i}]`;
    if (!isObject(raw)) {
      errors.push(`${at}: expected an object`);
      return;
    }

    for (const key of Object.keys(raw)) {
      if (!ROUTE_FIELDS.has(key)) {
        errors.push(`${at}.${key}: unknown field`);
      }
    }

    const id = raw.id;
    if (typeof id !== "string" || id.length === 0) {
      errors.push(`${at}.id: must be a non-empty string`);
      return;
    }
    if (seenIds.has(id)) {
      errors.push(`${at}.id: duplicate route id "${id}"`);
      return;
    }
    seenIds.add(id);

    const provider = raw.provider;
    if (typeof provider !== "string" || !knownProviderIds().includes(provider)) {
      errors.push(
        `${at}.provider: unknown provider ${JSON.stringify(provider)} (known: ${knownProviderIds().join(", ")})`,
      );
      return;
    }

    const model = raw.model;
    if (typeof model !== "string" || model.length === 0) {
      errors.push(`${at}.model: must be a non-empty string`);
      return;
    }

    const keyPool: string[] = [];
    if (raw.apiKey !== undefined) {
      if (typeof raw.apiKey !== "string" || raw.apiKey.length === 0) {
        errors.push(`${at}.apiKey: must be a non-empty string`);
        return;
      }
      keyPool.push(raw.apiKey);
    }
    if (raw.apiKeys !== undefined) {
      if (
        !Array.isArray(raw.apiKeys) ||
        raw.apiKeys.length === 0 ||
        !raw.apiKeys.every((k) => typeof k === "string" && k.length > 0)
      ) {
        errors.push(`${at}.apiKeys: must be a non-empty array of non-empty strings`);
        return;
      }
      keyPool.push(...raw.apiKeys);
    }
    if (keyPool.length === 0) {
      // Keyless routes are only valid for presets that declare no-auth
      // (local runtimes like ollama/vLLM).
      if (getPreset(provider)?.auth !== "none") {
        errors.push(`${at}: needs "apiKey" or "apiKeys"`);
        return;
      }
    }

    if (provider === "openai-compatible" && (typeof raw.baseUrl !== "string" || raw.baseUrl.length === 0)) {
      errors.push(`${at}.baseUrl: required when provider is "openai-compatible"`);
      return;
    }
    if (provider === "azure" && (typeof raw.baseUrl !== "string" || raw.baseUrl.length === 0)) {
      errors.push(`${at}.baseUrl: required when provider is "azure" (resource root, e.g. https://my-res.openai.azure.com)`);
      return;
    }
    if (provider === "azure") {
      if (typeof raw.apiVersion !== "string" || raw.apiVersion.length === 0) {
        errors.push(`${at}.apiVersion: required when provider is "azure" (e.g. "2024-10-21")`);
        return;
      }
    } else if (raw.apiVersion !== undefined) {
      errors.push(`${at}.apiVersion: only valid when provider is "azure"`);
      return;
    }
    for (const [field, ownerList, hint] of [
      ["region", ["bedrock", "vertex"], 'e.g. "us-east-1"'],
      ["project", ["vertex"], "GCP project id"],
    ] as const) {
      const owners: readonly string[] = ownerList;
      const value = raw[field];
      if (owners.includes(provider)) {
        if (typeof value !== "string" || value.length === 0) {
          errors.push(`${at}.${field}: required when provider is "${provider}" (${hint})`);
          return;
        }
      } else if (value !== undefined) {
        errors.push(`${at}.${field}: only valid when provider is ${owners.map((o) => `"${o}"`).join(" or ")}`);
        return;
      }
    }
    if (raw.baseUrl !== undefined && (typeof raw.baseUrl !== "string" || raw.baseUrl.length === 0)) {
      errors.push(`${at}.baseUrl: must be a non-empty string`);
      return;
    }

    if (raw.weight !== undefined && !(typeof raw.weight === "number" && raw.weight > 0)) {
      errors.push(`${at}.weight: must be a positive number`);
      return;
    }

    if (raw.headers !== undefined) {
      const h = raw.headers;
      if (
        !isObject(h) ||
        !Object.values(h).every((v) => typeof v === "string")
      ) {
        errors.push(`${at}.headers: must be an object of string -> string`);
        return;
      }
    }

    for (const [field, value] of [
      ["maxRetries", raw.maxRetries],
      ["timeoutMs", raw.timeoutMs],
      ["streamIdleTimeoutMs", raw.streamIdleTimeoutMs],
    ] as const) {
      if (value !== undefined && !positiveInt(value) && !(field === "maxRetries" && value === 0)) {
        errors.push(`${at}.${field}: must be a positive integer`);
        return;
      }
    }

    let limit: ModelRoute["limit"];
    if (raw.limit !== undefined) {
      if (!isObject(raw.limit)) {
        errors.push(`${at}.limit: expected an object`);
        return;
      }
      for (const key of Object.keys(raw.limit)) {
        if (!LIMIT_FIELDS.has(key)) {
          errors.push(`${at}.limit.${key}: unknown field (known: rpm, tpm)`);
        }
      }
      limit = {};
      if (raw.limit.rpm !== undefined) {
        if (!positiveInt(raw.limit.rpm)) {
          errors.push(`${at}.limit.rpm: must be a positive integer`);
          return;
        }
        limit.rpm = raw.limit.rpm as number;
      }
      if (raw.limit.tpm !== undefined) {
        if (!positiveInt(raw.limit.tpm)) {
          errors.push(`${at}.limit.tpm: must be a positive integer`);
          return;
        }
        limit.tpm = raw.limit.tpm as number;
      }
    }

    let budget: ModelRoute["budget"];
    if (raw.budget !== undefined) {
      if (!isObject(raw.budget)) {
        errors.push(`${at}.budget: expected an object`);
        return;
      }
      for (const key of Object.keys(raw.budget)) {
        if (!BUDGET_FIELDS.has(key)) {
          errors.push(`${at}.budget.${key}: unknown field (known: usd, windowMs)`);
        }
      }
      if (typeof raw.budget.usd !== "number" || !(raw.budget.usd > 0)) {
        errors.push(`${at}.budget.usd: must be a positive number`);
        return;
      }
      if (
        raw.budget.windowMs !== undefined &&
        !positiveInt(raw.budget.windowMs)
      ) {
        errors.push(`${at}.budget.windowMs: must be a positive integer`);
        return;
      }
      budget = { usd: raw.budget.usd as number, ...(raw.budget.windowMs !== undefined ? { windowMs: raw.budget.windowMs as number } : {}) };
    }

    routes.push({
      id,
      provider: provider as ModelRoute["provider"],
      model,
      apiKey: raw.apiKey as string | undefined,
      apiKeys: (raw.apiKeys as string[] | undefined)?.slice(),
      baseUrl: raw.baseUrl as string | undefined,
      apiVersion: raw.apiVersion as string | undefined,
      region: raw.region as string | undefined,
      project: raw.project as string | undefined,
      headers: raw.headers as Record<string, string> | undefined,
      maxRetries: raw.maxRetries as number | undefined,
      timeoutMs: raw.timeoutMs as number | undefined,
      streamIdleTimeoutMs: raw.streamIdleTimeoutMs as number | undefined,
      limit,
      budget,
      weight: raw.weight as number | undefined,
    });
  });

  if (errors.length > 0) {
    throw new ConfigError(errors.join("\n"));
  }

  return {
    routes,
    ...(typeof strategy === "string" ? { strategy: strategy as RouterConfig["strategy"] } : {}),
  };
}

File: packages/python/src/ai_router/config/parse.py
"""Hand-rolled configuration validation and interpolation (zero runtime dependencies).

Collects ALL violations and reports them with JSON-path style locations so both
the TypeScript and Python SDKs emit identical error messages.
"""

from __future__ import annotations

import os
import re
from collections.abc import Mapping
from typing import Any

from ..errors import ConfigError
from ..providers.presets import get_preset
from ..providers.registry import known_provider_ids
from ..routing.capabilities import ModelCapabilities
from .schema import (
    BudgetRule,
    LimitRule,
    ModelRoute,
    PolicyWeights,
    RouterConfig,
    RoutingStrategy,
)

ENV_PATTERN = re.compile(r"\$\{([A-Za-z_][A-Za-z0-9_]*)\}")

TOP_LEVEL_FIELDS = frozenset(["routes", "strategy", "weights"])
WEIGHT_FIELDS = frozenset(["cost", "speed", "reliability"])
ROUTE_FIELDS = frozenset(
    [
        "id",
        "provider",
        "model",
        "apiKey",
        "apiKeys",
        "baseUrl",
        "apiVersion",
        "region",
        "project",
        "headers",
        "maxRetries",
        "timeoutMs",
        "streamIdleTimeoutMs",
        "limit",
        "budget",
        "weight",
        "capabilities",
    ]
)
LIMIT_FIELDS = frozenset(["rpm", "tpm"])
BUDGET_FIELDS = frozenset(["usd", "windowMs"])
CAPABILITY_FIELDS = frozenset(
    [
        "streaming",
        "tools",
        "vision",
        "json",
        "structuredOutput",
        "reasoning",
        "audio",
        "embeddings",
        "multimodal",
        "longContext",
        "contextWindow",
    ]
)
STRATEGIES = frozenset(
    [
        "fallback",
        "round-robin",
        "weighted",
        "least-latency",
        "cheapest",
        "balanced",
        "quality-first",
    ]
)


def _interpolate_env(value: str, env: Mapping[str, str], path: str) -> str:
    if "${" not in value:
        return value

    def replace_var(match: re.Match[str]) -> str:
        var_name = match.group(1)
        if var_name not in env:
            raise ConfigError(f'{path}: environment variable "{var_name}" is not set')
        return env[var_name]

    return ENV_PATTERN.sub(replace_var, value)


def _interpolate_deep(obj: Any, env: Mapping[str, str], path: str) -> Any:
    if isinstance(obj, str):
        return _interpolate_env(obj, env, path)
    if isinstance(obj, list):
        return [_interpolate_deep(item, env, f"{path}[{i}]") for i, item in enumerate(obj)]
    if isinstance(obj, dict):
        return {k: _interpolate_deep(v, env, f"{path}.{k}") for k, v in obj.items()}
    return obj


def _is_dict(v: Any) -> bool:
    return isinstance(v, dict)


def _positive_int(v: Any) -> bool:
    return isinstance(v, int) and not isinstance(v, bool) and v > 0


def _positive_number(v: Any) -> bool:
    return (isinstance(v, (int, float)) and not isinstance(v, bool)) and v > 0


def parse_config(
    input_config: Any,
    env: Mapping[str, str] | None = None,
) -> RouterConfig:
    """Validates and parses a router config dictionary or JSON object with ${ENV_VAR} interpolation."""
    resolved_env = env if env is not None else os.environ
    interpolated = _interpolate_deep(input_config, resolved_env, "config")

    errors: list[str] = []

    if not _is_dict(interpolated):
        raise ConfigError("config: expected an object")

    routes_raw = interpolated.get("routes")
    if not isinstance(routes_raw, list):
        raise ConfigError("config.routes: expected an array")
    if len(routes_raw) == 0:
        raise ConfigError("config.routes: must not be empty")

    for key in interpolated.keys():
        if key not in TOP_LEVEL_FIELDS:
            errors.append(f"config.{key}: unknown field (known: routes, strategy, weights)")

    strategy_raw = interpolated.get("strategy")
    if strategy_raw is not None and (not isinstance(strategy_raw, str) or strategy_raw not in STRATEGIES):
        errors.append(
            'config.strategy: must be "fallback", "round-robin", "weighted", "least-latency", "cheapest", "balanced", or "quality-first"'
        )

    weights: PolicyWeights | None = None
    weights_raw = interpolated.get("weights")
    if weights_raw is not None:
        if not _is_dict(weights_raw):
            errors.append("config.weights: expected an object")
        else:
            for key in weights_raw.keys():
                if key not in WEIGHT_FIELDS:
                    errors.append(f"config.weights.{key}: unknown field (known: cost, speed, reliability)")

            w_dict: dict[str, float] = {}
            for key in WEIGHT_FIELDS:
                val = weights_raw.get(key)
                if val is None:
                    continue
                if not _positive_number(val):
                    errors.append(f"config.weights.{key}: must be a positive number")
                    continue
                w_dict[key] = float(val)

            if len(w_dict) == 0:
                errors.append("config.weights: expected at least one of cost, speed, reliability")
            elif len(errors) == 0:
                weights = PolicyWeights(
                    cost=w_dict.get("cost"),
                    speed=w_dict.get("speed"),
                    reliability=w_dict.get("reliability"),
                )

    seen_ids: set[str] = set()
    routes: list[ModelRoute] = []

    for i, raw in enumerate(routes_raw):
        at = f"config.routes[{i}]"
        if not _is_dict(raw):
            errors.append(f"{at}: expected an object")
            continue

        for key in raw.keys():
            if key not in ROUTE_FIELDS:
                errors.append(f"{at}.{key}: unknown field")

        route_id = raw.get("id")
        if not isinstance(route_id, str) or len(route_id) == 0:
            errors.append(f"{at}.id: must be a non-empty string")
            continue
        if route_id in seen_ids:
            errors.append(f'{at}.id: duplicate route id "{route_id}"')
            continue
        seen_ids.add(route_id)

        provider = raw.get("provider")
        known = known_provider_ids()
        if not isinstance(provider, str) or provider not in known:
            errors.append(f'{at}.provider: unknown provider "{provider}" (known: {", ".join(known)})')
            continue

        model = raw.get("model")
        if not isinstance(model, str) or len(model) == 0:
            errors.append(f"{at}.model: must be a non-empty string")
            continue

        key_pool: list[str] = []
        api_key = raw.get("apiKey")
        if api_key is not None:
            if not isinstance(api_key, str) or len(api_key) == 0:
                errors.append(f"{at}.apiKey: must be a non-empty string")
                continue
            key_pool.append(api_key)

        api_keys = raw.get("apiKeys")
        if api_keys is not None:
            if (
                not isinstance(api_keys, list)
                or len(api_keys) == 0
                or not all(isinstance(k, str) and len(k) > 0 for k in api_keys)
            ):
                errors.append(f"{at}.apiKeys: must be a non-empty array of non-empty strings")
                continue
            key_pool.extend(api_keys)

        if len(key_pool) == 0:
            preset = get_preset(provider)
            if not preset or preset.auth != "none":
                errors.append(f'{at}: needs "apiKey" or "apiKeys"')
                continue

        base_url = raw.get("baseUrl")
        if provider == "openai-compatible" and (not isinstance(base_url, str) or len(base_url) == 0):
            errors.append(f'{at}.baseUrl: required when provider is "openai-compatible"')
            continue
        if provider == "azure" and (not isinstance(base_url, str) or len(base_url) == 0):
            errors.append(
                f'{at}.baseUrl: required when provider is "azure" (resource root, e.g. https://my-res.openai.azure.com)'
            )

        api_version = raw.get("apiVersion")
        if provider == "azure":
            if not isinstance(api_version, str) or len(api_version) == 0:
                errors.append(f'{at}.apiVersion: required when provider is "azure" (e.g. "2024-10-21")')
        elif api_version is not None:
            errors.append(f'{at}.apiVersion: only valid when provider is "azure"')

        region = raw.get("region")
        if provider in ("bedrock", "vertex"):
            if not isinstance(region, str) or len(region) == 0:
                hint = 'e.g. "us-east-1"'
                errors.append(f'{at}.region: required when provider is "{provider}" ({hint})')
                continue
        elif region is not None:
            errors.append(f'{at}.region: only valid when provider is "bedrock" or "vertex"')
            continue

        project = raw.get("project")
        if provider == "vertex":
            if not isinstance(project, str) or len(project) == 0:
                errors.append(f'{at}.project: required when provider is "vertex" (GCP project id)')
                continue
        elif project is not None:
            errors.append(f'{at}.project: only valid when provider is "vertex"')
            continue

        if base_url is not None and (not isinstance(base_url, str) or len(base_url) == 0):
            errors.append(f"{at}.baseUrl: must be a non-empty string")
            continue

        weight = raw.get("weight")
        if weight is not None and not _positive_number(weight):
            errors.append(f"{at}.weight: must be a positive number")
            continue

        headers = raw.get("headers")
        if headers is not None:
            if not _is_dict(headers) or not all(isinstance(v, str) for v in headers.values()):
                errors.append(f"{at}.headers: must be an object of string -> string")
                continue

        max_retries = raw.get("maxRetries")
        if max_retries is not None and not (_positive_int(max_retries) or max_retries == 0):
            errors.append(f"{at}.maxRetries: must be a positive integer")
            continue

        timeout_ms = raw.get("timeoutMs")
        if timeout_ms is not None and not _positive_int(timeout_ms):
            errors.append(f"{at}.timeoutMs: must be a positive integer")
            continue

        stream_idle_timeout_ms = raw.get("streamIdleTimeoutMs")
        if stream_idle_timeout_ms is not None and not _positive_int(stream_idle_timeout_ms):
            errors.append(f"{at}.streamIdleTimeoutMs: must be a positive integer")
            continue

        limit: LimitRule | None = None
        limit_raw = raw.get("limit")
        if limit_raw is not None:
            if not _is_dict(limit_raw):
                errors.append(f"{at}.limit: expected an object")
                continue
            for k in limit_raw.keys():
                if k not in LIMIT_FIELDS:
                    errors.append(f"{at}.limit.{k}: unknown field (known: rpm, tpm)")
            rpm_val = limit_raw.get("rpm")
            if rpm_val is not None and not _positive_int(rpm_val):
                errors.append(f"{at}.limit.rpm: must be a positive integer")
                continue
            tpm_val = limit_raw.get("tpm")
            if tpm_val is not None and not _positive_int(tpm_val):
                errors.append(f"{at}.limit.tpm: must be a positive integer")
                continue
            limit = LimitRule(rpm=rpm_val, tpm=tpm_val)

        budget: BudgetRule | None = None
        budget_raw = raw.get("budget")
        if budget_raw is not None:
            if not _is_dict(budget_raw):
                errors.append(f"{at}.budget: expected an object")
                continue
            for k in budget_raw.keys():
                if k not in BUDGET_FIELDS:
                    errors.append(f"{at}.budget.{k}: unknown field (known: usd, windowMs)")
            usd_val = budget_raw.get("usd")
            if usd_val is None or not _positive_number(usd_val):
                errors.append(f"{at}.budget.usd: must be a positive number")
                continue
            win_ms = budget_raw.get("windowMs")
            if win_ms is not None and not _positive_int(win_ms):
                errors.append(f"{at}.budget.windowMs: must be a positive integer")
                continue
            budget = BudgetRule(usd=float(usd_val), windowMs=win_ms)

        capabilities: ModelCapabilities | None = None
        caps_raw = raw.get("capabilities")
        if caps_raw is not None:
            if not _is_dict(caps_raw):
                errors.append(f"{at}.capabilities: expected an object")
                continue
            for k in caps_raw.keys():
                if k not in CAPABILITY_FIELDS:
                    errors.append(
                        f"{at}.capabilities.{k}: unknown field (known: {', '.join(sorted(CAPABILITY_FIELDS))})"
                    )
            caps_dict: dict[str, Any] = {}
            for k in CAPABILITY_FIELDS:
                val = caps_raw.get(k)
                if val is None:
                    continue
                if k == "contextWindow":
                    if not _positive_int(val):
                        errors.append(f"{at}.capabilities.contextWindow: must be a positive integer")
                        break
                    caps_dict[k] = val
                else:
                    if not isinstance(val, bool):
                        errors.append(f"{at}.capabilities.{k}: must be a boolean")
                        break
                    caps_dict[k] = val
            capabilities = ModelCapabilities(**caps_dict)

        routes.append(
            ModelRoute(
                id=route_id,
                provider=provider,
                model=model,
                apiKey=api_key,
                apiKeys=list(api_keys) if api_keys is not None else None,
                baseUrl=base_url,
                apiVersion=api_version,
                region=region,
                project=project,
                headers=dict(headers) if headers is not None else None,
                maxRetries=max_retries,
                timeoutMs=timeout_ms,
                streamIdleTimeoutMs=stream_idle_timeout_ms,
                limit=limit,
                budget=budget,
                weight=float(weight) if weight is not None else None,
                capabilities=capabilities,
            )
        )

    if len(errors) > 0:
        raise ConfigError("\n".join(errors))

    from typing import cast

    return RouterConfig(
        routes=routes,
        strategy=cast(RoutingStrategy, strategy_raw) if isinstance(strategy_raw, str) else None,
        weights=weights,
    )
